### PR TITLE
Connect-DbaInstance - sql.connection.experimental to be able to safely test new code paths

### PR DESCRIPTION
**Pull request by andreasjordan**
In order to be able to test new functions in various environments, the switch "experimental" is introduced. This switch can be set with "Set-DbatoolsConfig -FullName sql.connection.experimental -Value $true" for the active session and within this function leads to the following code path being used.

Please have a look and give feedback. I will continue to work on this and publish code to test this as well in the next days.

### CODE REVIEW

**Comment by wsmelton** [comment]
So will `Connect-SqlInstance` be adjusted to handle calling this as well or is this purely for those wanting to just call `Connect-DbaInstance`?

The only command that publically calls `Connect-DbaInstance` is a few of the Invoke commands. I know `Invoke-DbaQuery` and recall there is one other one but all our functions call `Connect-SqlInstance`.

So it should be clearly stated which path will be supported if our internal functions at this time would not utilize it.

**Comment by andreasjordan** [comment]
@wsmelton - thanks a lot, haven't had Connect-SqlInstance in my mind because I always use Connect-DbaInstance.

Just had a quick look: Yes, it need attention:
```
if ($AzureUnsupported -and $server.DatabaseEngineType -eq "SqlAzureDatabase") {
```
This will only work if $server is set somewhere beforehand - as i is not set in this command.

If you could tell me something about the intention of this command, that would be great.

**Comment by wsmelton** [comment]
We started out with just having `Connect-SqlInstance` used for internal and created a copy as `Copy-DbaInstance`. After some time we then switched to just having `Connect-SqlInstance` be a wrapper to calling `Copy-DbaInstance` for easier maintenance and all.

I'm not sure but since the only real use for it now is dealing with internal for Azure if maybe it's time to get rid of it? @potatoqualitee @niphlod

**Comment by wsmelton** [comment]
Something we can discuss in an issue if desired about this change/PR.

**Comment by andreasjordan** [comment]
I've worked on New-DbaConnectionString because it is used in the Azure part of Connect-DbaInstance.
And I've used the feature of Connect-DbaInstance to setup a server object from a connection string.
Positiv: Works very well and connection pooling should work, because the server smo handles the connection.
Negative: The property ConnectionContext.ConnectionString has all the details of the connection, but these are not transfered to the individual properties of ConnectionContext. So $server.ConnectionContext.ApplicationName is still 'SQL Management' - but the value from ConnectionString is definitly used. 
I had a look at all the code: Only the test for Connect-DbaInstance tests for the individual properties.
And in dbatools.Types.ps1xml there is $this.ConnectionContext.DatabaseName referenced - but I don't know what this file is for.
So: Can use this as a general way? So parameter -> connection string (with New-DbaConnectionString) -> server smo -> connect.
Alternative: parameter -> connection string (with New-DbaConnectionString) -> empty smo -> set individual properties of ConnectionContext from connection string -> connect.

**Comment by niphlod** [comment]
+1 for having one only function to connect. 
Dealing with commands using -DbaInstance vs -Sqlinstance should be just a matter of choice, since -SqlInstance is an alias.

On the "properties not set" .... we *could* always set some of the "important" ones manually. DatabaseName e.g. is used in Invoke-DbaQuery to see if the targeted database is the correct one or not.

**Comment by andreasjordan** [comment]
Hi @niphlod :
sorry to say, updating the properties is only possibe bevor connecting:

```
# Can we update the individual properties from the connection string?
$connStringBuilder = New-Object -TypeName System.Data.SqlClient.SqlConnectionStringBuilder -ArgumentList $server.ConnectionContext.ConnectionString
# Sorry, not allowed:
$server.ConnectionContext.ApplicationName = $connStringBuilder['Application Name']
$server.ConnectionContext.DatabaseName = $connStringBuilder['Initial Catalog']
```

Here is my test-code: https://github.com/andreasjordan/demos/blob/master/dbatools/Test_Experimental_Connect-DbaInstance.ps1
(Last line is not pushed jet - just tested, but DatabaseName is not settable.)
But the current Database is available in `$server.ConnectionContext.SqlConnectionObject.Database`.

**Comment by niphlod** [comment]
well, in the need we can check for $server.ConnectionContext.SqlConnectionObject.Database so no real problems there

**Comment by andreasjordan** [comment]
Here the DatabaseName from the ConnectionContext is used:

https://github.com/sqlcollaborative/dbatools/blob/05fbf9898caf3922d1c9bf8017465bd57e5d387a/functions/Invoke-DbaQuery.ps1#L351-L355

So even today, there might be issues when using a server object that was created from a connection string.
Will include Invoke-DbaQuery in my ideas and tests...

**Comment by potatoqualitee** [comment]
Yes, the `Connect-SqlInstance` is a wrapper that allows us to easily show that a command does not support azure. It also gives us greater flexibility if we have to change anything in all the commands in the future. Not sure I see a benefit to switching to Dba since this is just a wrapper.

**Comment by potatoqualitee** [comment]
Also, thanks so much for the docs and changes! 🙇  Looks great!

**Comment by andreasjordan** [comment]
Sometimes it is very frustrating. The approach with the connection string probably does not work at all:

```
$instanceName = 'SRV1\SQL2016'

'Test 1:'
1..5 | ForEach-Object -Process {
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $instanceName
    $server.ConnectionContext.ProcessID
}

'Test 2:'
1..5 | ForEach-Object -Process {
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $instanceName
    $server.ConnectionContext.ApplicationName = 'Test'
    $server.ConnectionContext.ProcessID
}

'Test 3:'
1..5 | ForEach-Object -Process {
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $instanceName
    $server.ConnectionContext.ConnectionString = "Data Source=$instanceName;Integrated Security=True;MultipleActiveResultSets=False;Encrypt=False;TrustServerCertificate=False;Application Name=Test"
    $server.ConnectionContext.ProcessID
}
```

The result ist:

```
Test 1:
123
123
123
123
123
Test 2:
136
136
136
136
136
Test 3:
136
139
140
145
144
```

The first connection in the last block reuses the previous connection, but then new connections are opened.
So building just every connection from a connection string looks good in the code, but does not work. 

Or am I missing something?

**Comment by potatoqualitee** [comment]
that's been my experience @andreasjordan - i'm trying to follow what's recommended by Microsoft but the resulting connections do not work as expected. In addition to being pretty, the connectionstring also allows us to be flexible with all of the different ways we want to allow people to authenticate. Perhaps we can reach out to @shueybubbles to get clarification on the best way to create a connection using SMO.

Does `Get-SqlInstance` (or whichever is the equivalent command from `SqlServer`) work like you expect?

**Comment by potatoqualitee** [comment]
hey @niphlod - I keep getting the failure

```
Describe : dbatools style
Context  : formatting
Name     : It Connect-DbaInstance.ps1 is not compliant with the OTBS formatting style. Please run Invoke-DbatoolsFormatter against the failing file and commit the changes.
Result   : Failed
Message  : Expected 0, but got 1.
```

I ran it against ` Invoke-DbatoolsFormatter`, which took a while but produced no changes. Then I compared the encoding, no issues. Any idea what's causing this failure?

**Comment by shueybubbles** [comment]
I invite you to open an issue against SMO: https://github.com/microsoft/sqlmanagementobjects/issues

If you are trying to force Server to use pooled connections, try setting `$server.ConnectionContext.NonPooledConnection = false` before setting ConnectionString. 

I'd be curious to know if constructing your own `ServerConnection` instead of relying on Server creating one for you produces better results. 

You can construct a `SqlConnectionInfo` object (yes it's ridiculously redundant with `SqlConnectionStringBuilder`), set a bunch of properties on it, then pass it to the `ServerConnection` constructor, then pass the `ServerConnection` to the Server constructor.

Or you can hand a `SqlConnection` to the `ServerConnection` constructor for ultimate control.

We have a regression test that uses SqlConnectionInfo
```
                    var tableCreated = db.CreateTable("DropMe");
                    var ci = new SqlConnectionInfo
                    {
                        ServerName = this.SqlConnectionStringBuilder.DataSource,
                        UseIntegratedSecurity = this.SqlConnectionStringBuilder.IntegratedSecurity,
                        Pooled = true
                    };
                    if (!ci.UseIntegratedSecurity)
                    {
                        ci.UserName = this.SqlConnectionStringBuilder.UserID;
                        ci.Password = this.SqlConnectionStringBuilder.Password;
                    }
                    var smoServer = new _SMO.Server(new ServerConnection(ci));
                    var tableCollectionEntry = smoServer.Databases[db.Name].Tables[tableCreated.Name];
                    Assert.That(tableCollectionEntry.RowCount, _NU.Is.EqualTo(0), "RowCount");
```

**Comment by potatoqualitee** [comment]
Heck yes, thank you so much for the feedback @shueybubbles ! I'll get in touch with @andreasjordan to see what we can figure out.

**Comment by andreasjordan** [comment]
Using SqlConnectionInfo and ServerConnection works!

```
$instanceName = 'SRV1\SQL2016'

'Test 1:'
1..5 | ForEach-Object -Process {
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $instanceName
    $server.ConnectionContext.ProcessID
}

'Test 2:'
1..5 | ForEach-Object -Process {
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $instanceName
    $server.ConnectionContext.ApplicationName = 'Test'
    $server.ConnectionContext.ProcessID
}

'Test 3:'
1..5 | ForEach-Object -Process {
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $instanceName
    $server.ConnectionContext.NonPooledConnection = $false  # This doesn't help
    $server.ConnectionContext.ConnectionString = "Data Source=$instanceName;Integrated Security=True;MultipleActiveResultSets=False;Encrypt=False;TrustServerCertificate=False;Application Name=Test"
    $server.ConnectionContext.ProcessID
}

'Test 4:'
1..5 | ForEach-Object -Process {
    $connInfo = New-Object Microsoft.SqlServer.Management.Common.SqlConnectionInfo $instanceName
    $connInfo.ApplicationName = 'Test'
    $srvConn = New-Object Microsoft.SqlServer.Management.Common.ServerConnection $connInfo
    $server = New-Object Microsoft.SqlServer.Management.Smo.Server $srvConn
    $server.ConnectionContext.ProcessID
}
```

image


Next steps: See if it works with more properties, sql authentication and azure...

Thanks @shueybubbles for the input.

**Comment by shueybubbles** [comment]
be careful about using processid on Azure, it can change at any time, as the client hides TDS-level disconnects/reconnects from you.

**Comment by andreasjordan** [comment]
When using SqlConnectionInfo, we have a problem with these parameters of Connect-DbaInstance because they are not properties of SqlConnectionInfo and can only be set via a connection string:
* FailoverPartner
* MultiSubnetFailover
* AppendConnectionString

So we have to build a new server smo from a connection string if one of there parameters is used.
Or do you have other ideas?

**Comment by shueybubbles** [comment]
you can put whatever you want in `SqlConnectionInfo.AdditionalParameters`, that string gets appended as-is to the connection string.

**Comment by andreasjordan** [comment]
Ok, but how to set MultipleActiveResultSets?

```
$connInfo = New-Object Microsoft.SqlServer.Management.Common.SqlConnectionInfo $instanceName
$connInfo.AdditionalParameters = 'MultipleActiveResultSets=True'
$connInfo.ConnectionString
```
results in
```
Data Source=SRV1\SQL2016;Integrated Security=True;MultipleActiveResultSets=False;Encrypt=False;TrustServerCertificate=False;MultipleActiveResultSets=True
```

**Comment by andreasjordan** [comment]
Oh sorry, I can set that later in the ConnectionContext - ignore tha last comment...

**Comment by andreasjordan** [comment]
Code:

```
$instanceName = 'SRV1\SQL2016'

'Test 1:'
1..3 | ForEach-Object -Process {
    $server = Connect-DbaInstance -SqlInstance $instanceName
    $server.ConnectionContext.ProcessID
}

'Test 2:'
1..3 | ForEach-Object -Process {
    $server = Connect-DbaInstance -SqlInstance $instanceName
    $server.ConnectionContext.ProcessID
    $server = Connect-DbaInstance -SqlInstance $instanceName -Database tempdb
    $server.ConnectionContext.ProcessID
}

'Test 3:'
1..3 | ForEach-Object -Process {
    Invoke-DbaQuery -SqlInstance $instanceName -Query 'SELECT @@SPID' -As SingleValue
}

'Test 4:'
1..3 | ForEach-Object -Process {
    Invoke-DbaQuery -SqlInstance $instanceName -Query 'SELECT @@SPID' -As SingleValue
    Invoke-DbaQuery -SqlInstance $instanceName -Database tempdb -Query 'SELECT @@SPID' -As SingleValue
}

'Test 5:'
$server = Connect-DbaInstance -SqlInstance $instanceName
1..3 | ForEach-Object -Process {
    Invoke-DbaQuery -SqlInstance $server -Query 'SELECT @@SPID' -As SingleValue
    Invoke-DbaQuery -SqlInstance $server -Database tempdb -Query 'SELECT @@SPID' -As SingleValue
}
```

Result:

```
Test 1:
63
63
63
Test 2:
63
62
63
62
63
62
Test 3:
63
63
63
Test 4:
63
62
63
62
63
62
Test 5:
77
71
77
71
77
71
```

Looks good.

Will continue tests in the next days...

**Comment by andreasjordan** [comment]
Updated the authentication part and tested the different methods (test code: https://github.com/andreasjordan/demos/blob/master/dbatools/Test_Experimental_Connect-DbaInstance.ps1).

There are still al lot of TODO in the code, some with questions - so you are invited to have a look and give feedback.

**Comment by andreasjordan** [comment]
I have forced the new code to let AppVeyor run all tests - and they passed!

**Comment by andreasjordan** [comment]
@shueybubbles - I would be very happy if you could have a look at the code.

**Comment by potatoqualitee** [comment]
updated from dev with code that should help that failing test 👍

**Comment by potatoqualitee** [comment]
@jpomfret would you have time to test out Connect-DbaInstance in this branch? I believe you've been impacted by ConnectionPooling in the past.

**Comment by jpomfret** [comment]
Morning folks,

Pulled down this branch and tested out the connection pooling using this:

```
$server = Connect-DbaInstance -SqlInstance mssql1
Invoke-DbaQuery -SqlInstance $server -Query 'select 1'
Get-DbaDbTable -SqlInstance $server -Database master
```

Confirmed with Xevents it is all coming in on the same spid
image

That being said - I'm not sure I'm totally following the addition of the experimental config. Perhaps adding an example in the function would be good for this?

File: functions/Connect-DbaInstance.ps1 [hunk 1]
```diff
@@ -409,6 +421,428 @@ function Connect-DbaInstance {
...
+            This switch can be set with "Set-DbatoolsConfig -FullName sql.connection.experimental -Value $true" for the active session
+            and within this function leads to the following code path being used.
+            All the sub paths inside the following if clause will end with a continue, so the normal code path is not used.
+            #>
+            if (Get-DbatoolsConfigValue -FullName sql.connection.experimental) {
+                <#
+                Best practice:
+                * Create a smo server object by submitting the name of the instance as a string to SqlInstance and additional parameters to configure the connection
+                * Reuse the smo server object in all following calls as SqlInstance
+                * When reusing the smo server object, only the following additial parameters are allowed with Connect-DbaInstance:
```
Comment (jpomfret) [inline functions/Connect-DbaInstance.ps1:436]:
```suggestion
                * When reusing the smo server object, only the following additional parameters are allowed with Connect-DbaInstance:
```

File: functions/Connect-DbaInstance.ps1 [hunk 2]
```diff
@@ -409,6 +421,428 @@ function Connect-DbaInstance {
...
+                  - Database (command clones ConnectionContext and returns new smo server object)
+                  - AzureUnsupported (command failes if target is Azure)
```
Comment (jpomfret) [inline functions/Connect-DbaInstance.ps1:438]:
```suggestion
                  - AzureUnsupported (command fails if target is Azure)
```

File: functions/Connect-DbaInstance.ps1 [hunk 3]
```diff
@@ -409,6 +421,428 @@ function Connect-DbaInstance {
...
+                  - MinimumVersion (command failes if target version is too old)
```
Comment (jpomfret) [inline functions/Connect-DbaInstance.ps1:439]:
```suggestion
                  - MinimumVersion (command fails if target version is too old)
```

**Review by jpomfret — COMMENTED** [review]
couple of typos in the help :)

**Comment by andreasjordan** [comment]
@jpomfret : Thanks for the review. I fixed the typos and added (a first version) of an example.

**Comment by andreasjordan** [comment]
Does anyone know why a lot of tests are failing?

**Comment by jpomfret** [comment]
Looks like there is actual issues, failing for me locally too.

Tested these two and get failures:
- Get-DbaDbIdentity.Tests.ps1
- Get-DbaDbOrphanUser.Tests.ps1

Looking at Get-DbaDbIdentity tests this is in the BeforeAll and fails 

```
$db = Get-DbaDatabase -SqlInstance $script:instance1 -Database tempdb
$db.Query("CREATE TABLE dbo.dbatoolsci_example (Id int NOT NULL IDENTITY (125, 1), Value varchar(5));
INSERT INTO dbo.dbatoolsci_example(Value) Select 1;
CREATE TABLE dbo.dbatoolsci_example2 (Id int NOT NULL IDENTITY (5, 1), Value varchar(5));
INSERT INTO dbo.dbatoolsci_example2(Value) Select 1;")
```

Looks like a connection issue... ? 
```
Exception calling "Query" with "1" argument(s): "Exception calling "ExecuteWithResults" with "1" argument(s): "Execute with results failed for Database 'tempdb'. ""
At line:1 char:1
+ $db.Query("CREATE TABLE dbo.dbatoolsci_example (Id int NOT NULL IDENT ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ScriptMethodRuntimeException
```

Does it run successfully for you?
```
Invoke-Pester .\tests\Get-DbaDbOrphanUser.Tests.ps1
```

**Comment by andreasjordan** [comment]
I don't have a 2008R2 in my lab and I don't know if I set up the test environment correct.

I get:

image

**Comment by jpomfret** [comment]
I was running against a 2017 container instance.

It looks like you might have Pester v5.0+ installed. Can you try it with Pester 4.x.x?

**Comment by andreasjordan** [comment]
image

**Comment by jpomfret** [comment]
That's fine - is that the one imported in your session if you do Get-Module ?

**Comment by andreasjordan** [comment]
Just installed 4.10.1, now I get:

image

**Comment by potatoqualitee** [comment]
trying to run them again

**Comment by potatoqualitee** [comment]
Tests fail on unrelated command

**Comment by potatoqualitee** [comment]
Thanks again everyone who helped and provided feedback 💯 ❤️

**Pull request merged**